src/Microsoft.Azure.ServiceBus/Core/MessageBatch.cs

```diff
@@ -0,0 +1,141 @@
+
+        /// <summary>
+        /// Construct a new batch with a maximum batch size and outgoing plugins callback.
+        /// <remarks>
+        /// To construct a batch at run-time, use <see cref="MessageSender"/>, <see cref="QueueClient"/>, or <see cref="TopicClient"/>.
+        /// Use this constructor for testing and custom implementations.
+        /// </remarks>
+        /// </summary>
+        /// <param name="maximumBatchSize">Maximum batch size allowed for batch.</param>
+        /// <param name="pluginsCallback">Plugins callback to invoke on outgoing messages regisered with batch.</param>
+        internal MessageBatch(ulong maximumBatchSize, Func<Message, Task<Message>> pluginsCallback)
+        {
+            this.maximumBatchSize = maximumBatchSize;
+            this.pluginsCallback = pluginsCallback;
+            this.datas = new List<Data>();
+            this.result = AmqpMessage.Create(datas);
+        }
+
+        /// <summary>
+        /// Add <see cref="Message"/> to the batch if the overall size of the batch with the added message is not exceeding the batch maximum.
+        /// </summary>
+        /// <param name="message"><see cref="Message"/> to add to the batch.</param>
+        /// <returns></returns>
+        public async Task<bool> TryAdd(Message message)
+        {
+            ThrowIfDisposed();
+
+            message.VerifyMessageIsNotPreviouslyReceived();
+
+            var processedMessage = await pluginsCallback(message);
+
+            var amqpMessage = AmqpMessageConverter.SBMessageToAmqpMessage(processedMessage);
+
+            if (firstMessage == null)
+            {
+                firstMessage = amqpMessage;
+
+                if (processedMessage.MessageId != null)
+                {
+                    result.Properties.MessageId = processedMessage.MessageId;
+                }
+
+                if (processedMessage.SessionId != null)
+                {
+                    result.Properties.GroupId = processedMessage.SessionId;
+                }
+
+                if (processedMessage.PartitionKey != null)
+                {
+                    result.MessageAnnotations.Map[AmqpMessageConverter.PartitionKeyName] = processedMessage.PartitionKey;
+                }
+
+                if (processedMessage.ViaPartitionKey != null)
+                {
+                    result.MessageAnnotations.Map[AmqpMessageConverter.ViaPartitionKeyName] = processedMessage.ViaPartitionKey;
+                }
+            }
+
+            var data = AmqpMessageConverter.ToData(amqpMessage);
+            datas.Add(data);
+
+            if (Size <= maximumBatchSize)
+            {
+                return true;
+            }
+
+            datas.Remove(data);
+            return false;
+
+        }
+
+        /// <summary>
+        /// Number of messages in batch.
+        /// </summary>
+        public int Length => datas.Count;
+
+        internal ulong Size => (ulong) result.SerializedMessageSize;
```

Comment:
@nemakam I'm contemplating to make this public.
Reason: to be able to know what batch size was sent.